zetaclient/bitcoin/bitcoin_client_test.go

```diff
@@ -26,3 +37,131 @@ func TestConfirmationThreshold(t *testing.T) {
+		invalidVb := blockVb
+		invalidVb.Weight = blockVb.Tx[0].Weight - 1
+		_, err := CalcBlockAvgFeeRate(&invalidVb, &chaincfg.MainNetParams)
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "less than coinbase tx weight"))
+	})
+	t.Run("invalid block height should fail", func(t *testing.T) {
+		invalidVb := blockVb
+		invalidVb.Height = 0
+		_, err := CalcBlockAvgFeeRate(&invalidVb, &chaincfg.MainNetParams)
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "invalid block height"))
+
+		invalidVb.Height = math.MaxInt32 + 1
+		_, err = CalcBlockAvgFeeRate(&invalidVb, &chaincfg.MainNetParams)
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "invalid block height"))
+	})
+	t.Run("failed to decode coinbase tx", func(t *testing.T) {
+		invalidVb := blockVb
+		invalidVb.Tx = []btcjson.TxRawResult{blockVb.Tx[0], blockVb.Tx[1]}
+		invalidVb.Tx[0].Hex = "invalid hex"
+		_, err := CalcBlockAvgFeeRate(&invalidVb, &chaincfg.MainNetParams)
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "failed to decode coinbase tx"))
+	})
+	t.Run("1st tx is not coinbase", func(t *testing.T) {
+		invalidVb := blockVb
+		invalidVb.Tx = []btcjson.TxRawResult{blockVb.Tx[1], blockVb.Tx[0]}
+		_, err := CalcBlockAvgFeeRate(&invalidVb, &chaincfg.MainNetParams)
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "not coinbase tx"))
+	})
+	t.Run("miner earned less than subsidy", func(t *testing.T) {
+		invalidVb := blockVb
+		coinbaseTxBytes, err := hex.DecodeString(blockVb.Tx[0].Hex)
+		require.NoError(t, err)
+		coinbaseTx, err := btcutil.NewTxFromBytes(coinbaseTxBytes)
+		require.NoError(t, err)
+		msgTx := coinbaseTx.MsgTx()
+
+		// reduce subsidy by 1 satoshi
+		for i := range msgTx.TxOut {
+			if i == 0 {
+				msgTx.TxOut[i].Value = blockchain.CalcBlockSubsidy(int32(blockVb.Height), &chaincfg.MainNetParams) - 1
+			} else {
+				msgTx.TxOut[i].Value = 0
+			}
+		}
+		// calculate fee rate on modified coinbase tx
+		var buf bytes.Buffer
+		err = msgTx.Serialize(&buf)
+		require.NoError(t, err)
+		invalidVb.Tx[0].Hex = hex.EncodeToString(buf.Bytes())
+		_, err = CalcBlockAvgFeeRate(&invalidVb, &chaincfg.MainNetParams)
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "less than subsidy"))
+	})
+}
+
+func TestCalcDepositorFee828440(t *testing.T) {
+	// load archived block 828440
+	var blockVb btcjson.GetBlockVerboseTxResult
+	err := testutils.LoadObjectFromJSONFile(&blockVb, path.Join(testutils.TestDataPath, "bitcoin_block_trimmed_828440.json"))
+	require.NoError(t, err)
+
+	// should return default fee if it's a regtest block
+	fee := CalcDepositorFee(&blockVb, 18444, &chaincfg.RegressionNetParams, log.Logger)
+	require.Equal(t, DefaultDepositorFee, fee)
+
+	// should return dynamic fee if it's a testnet block
+	fee = CalcDepositorFee(&blockVb, 18332, &chaincfg.TestNet3Params, log.Logger)
+	require.NotEqual(t, DefaultDepositorFee, fee)
```

Comment:
Can't we be more precise? It should be consistent since we use a static file?

Reply:
https://github.com/zeta-chain/node/pull/1747/commits/2124f357744661b6646c324d2cb090cb27f844c1